File: 01 CLASE ENTERO/js/Entero.js

class Entero {
    //Atributos 
    Num;
    //Construtor 
    constructor(Numero) {
        this.Num = Numero;
    }

    // Class ---> Cargar Numero

    setNum() {
        const input = document.getElementById("enteroInput"); 
        if (input.value == ""){
            return 0; 
        }
        this.Num = parseInt(input.value, 10); 
      
    }

    //Class ---> Mostrar el valor de numero 
    getNum() {
        return this.Num;

    }



    //Class ---> Obtener el valor de Numero   
    showNum() {
        const resultadoDiv = document.getElementById("resultado")
        resultadoDiv.textContent = this.getNum();

    }

    showResultado(respuesta){
        const resultadoDiv = document.getElementById("resultado"); 
        resultadoDiv.textContent = respuesta; 
    }

    IngrementarNum(){
        this.Num = this.Num + 1 ;
    }

    DecrementarNum(){
        this.Num = this.Num - 1 ;
    }


    esParOesInpar(){
        return (this.Num % 2 ==0) 
    }

    esNegaesPosi(){
        return(this.Num > 0 )
    }

    esNumeroPerfecto() {
        let suma = 0;
        // Sumamos los divisores propios del número
        for (let i = 1; i < this.Num; i++) {
            if (this.Num % i === 0) {
                suma += i;
            }
        }
        // Retornamos true si la suma de los divisores es igual al número
        return suma === this.Num;
    }

    esNumeroPrimo() {
        if (this.Num < 2) return false; 
        for (let i = 2; i <= Math.sqrt(this.Num); i++) {
            if (this.Num % i === 0) {
                return false;
            }
        }
        return true;
    }

    fibonacci() {
        let a = 0, b = 1;
        while (a <= this.Num) {
            if (a === this.Num) {
                return true;
            }
            [a, b] = [b, a + b];
        }
        return false;
    }

    Generarfibonacci() {
        let serie = [0, 1];
        for (let i = 2; i < this.Num; i++) {
            serie.push(serie[i - 1] + serie[i - 2]);
        }
        return serie.slice(0, this.Num);
    }

    esArmstrong() {
        let numStr = this.Num.toString(); 
        let numDigitos = numStr.length; 
        let suma = 0;
        
        for (let i = 0; i < numDigitos; i++) {
            suma += Math.pow(parseInt(numStr[i]), numDigitos); 
        }

        return suma === this.Num;
    }

    esPerfectoGeneralizado(k) {
        let suma = 0;
        for (let i = 1; i < this.Num; i++) {
            if (this.Num % i === 0) {
                suma += Math.pow(i, k);
            }
        }
        return suma === this.Num;
    }

    calcularFilaPascal() {
        const fila = [];
        for (let k = 0; k <= this.Num; k++) {
            fila.push(this.factorial(this.Num) / (this.factorial(k) * this.factorial(this.Num - k)));
        }
        return fila;
    }

    factorial(n) {
        if (n === 0 || n === 1) {
            return 1;
        }
        let resultado = 1;
        for (let i = 2; i <= n; i++) {
            resultado *= i;
        }
        return resultado;
    }

    cifrarCesar(desplazamiento) {
        const numStr = this.Num.toString(); // Convertir el número a string
        let resultado = '';

        for (let i = 0; i < numStr.length; i++) {
            const digito = parseInt(numStr[i], 10); // Obtener cada dígito
            const nuevoDigito = (digito + desplazamiento) % 10; // Desplazar dentro del rango 0-9
            resultado += nuevoDigito.toString(); // Concatenar el nuevo dígito
        }

        return resultado; // Devolver el número cifrado como string
    }

     tienePrimoGemelo() {
        if (!this.esNumeroPrimo()) return false;

        const esPrimo = (n) => {
            if (n < 2) return false;
            for (let i = 2; i <= Math.sqrt(n); i++) {
                if (n % i === 0) return false;
            }
            return true;
        };

        return esPrimo(this.Num - 2) || esPrimo(this.Num + 2);
    }

    descomponerEnFactoresPrimos() {
        let n = this.Num;
        const factores = {};
        let divisor = 2;

        while (n > 1) {
            let potencia = 0;
            while (n % divisor === 0) {
                n /= divisor;
                potencia++;
            }
            if (potencia > 0) {
                factores[divisor] = potencia;
            }
            divisor++;
        }
        return factores;
    }

    calcularCatalan() {
        const factorial = (n) => {
            return n === 0 ? 1 : n * factorial(n - 1);
        };

        const catalan = factorial(2 * this.Num) / (factorial(this.Num + 1) * factorial(this.Num));
        return catalan;
    }

    esMersenne() {
        const log2p1 = Math.log2(this.Num + 1);
        return Number.isInteger(log2p1) && this.esNumeroPrimo(log2p1);
    }
}

//Las Funciones = Button HTML 
var ClaseEntero = new Entero(0);

//CREAR EL OBJETO
function cargarNum() {
    ClaseEntero.setNum();
    alert("ok")
}

//Mostrar el Objeto 

function MostrarNum() {
    ClaseEntero.showNum();
}

function IngrementarNumero(){
    ClaseEntero.IngrementarNum(); 
    ClaseEntero.showNum(); 
}

function DecrementarNumero(){
    ClaseEntero.DecrementarNum();
    ClaseEntero.showNum(); 
}

function esParImparNum(){
    var respuesta = ClaseEntero.esParOesInpar(); 
    var resp = respuesta ? "Es Num Par":"Es Num Impar"
    ClaseEntero.showResultado(resp);
}

function PositiveNegative(){
    var respuesta = ClaseEntero.esNegaesPosi(); 
    var resp = respuesta ? "Es Positivo":"Es Negativo"
    ClaseEntero.showResultado(resp); 
}

function esNumeroPerfecto() {
    var respuesta = ClaseEntero.esNumeroPerfecto();
    var resp = respuesta ? "Es un número perfecto" : "No es un número perfecto";
    ClaseEntero.showResultado(resp);
}

function esNumeroPrimo() {
    var respuesta = ClaseEntero.esNumeroPrimo();
    var resp = respuesta ? "Es un número primo" : "No es un número primo";
    ClaseEntero.showResultado(resp);
}

function PillarFibonacci() {
    var pertenece = ClaseEntero.fibonacci();
    var resp = pertenece ? `El número ${ClaseEntero.getNum()} pertenece a la serie de Fibonacci.` : `Error: El número ${ClaseEntero.getNum()} no pertenece a la serie de Fibonacci.`;
    ClaseEntero.showResultado(resp);
}

function generarFibonacci() {
    var serie = ClaseEntero.Generarfibonacci();
    var resp = "Serie de Fibonacci: " + serie.join(", ");
    ClaseEntero.showResultado(resp);
}

function esNumeroArmstrong() { 
    var respuesta = ClaseEntero.esArmstrong();
    var resp = respuesta ? "Es un número de Armstrong" : "No es un número de Armstrong";
    ClaseEntero.showResultado(resp);
}

function esNumeroPerfectoGeneralizado() {
    const k = parseInt(prompt("Ingrese el valor de k: "), 10);
    var respuesta = ClaseEntero.esPerfectoGeneralizado(k);
    var resp = respuesta ? `Es un número perfecto generalizado con respecto a k=${k}` : `No es un número perfecto generalizado con respecto a k=${k}`;
    ClaseEntero.showResultado(resp);
}

function mostrarFilaPascal() {
    const filaPascal = ClaseEntero.calcularFilaPascal();
    var resp = `Fila ${ClaseEntero.getNum()} del triángulo de Pascal: [${filaPascal.join(", ")}]`;
    ClaseEntero.showResultado(resp);
}

function mostrarCifradoCesar() {
    const desplazamiento = parseInt(prompt("Ingrese el valor del desplazamiento: "), 10);
    const resultadoCifrado = ClaseEntero.cifrarCesar(desplazamiento);
    var resp = `Número cifrado con desplazamiento ${desplazamiento}: ${resultadoCifrado}`;
    ClaseEntero.showResultado(resp);
}

function mostrarPrimoGemelo() {
    var resp = ClaseEntero.tienePrimoGemelo() ? `El número ${ClaseEntero.getNum()} tiene un primo gemelo.` : `El número ${ClaseEntero.getNum()} no tiene un primo gemelo.`;
    ClaseEntero.showResultado(resp);
}

function mostrarFactoresPrimos() {
    const factores = ClaseEntero.descomponerEnFactoresPrimos();
    var resp = `Descomposición en factores primos: ${JSON.stringify(factores)}`;
    ClaseEntero.showResultado(resp);
}

function mostrarCatalan() {
    var catalan = ClaseEntero.calcularCatalan();
    var resp = `El número de Catalan para ${ClaseEntero.getNum()} es: ${catalan}`;
    ClaseEntero.showResultado(resp);
}

function mostrarMersenne() {
    var resp = ClaseEntero.esMersenne() ? `El número ${ClaseEntero.getNum()} es un número de Mersenne.` : `El número ${ClaseEntero.getNum()} no es un número de Mersenne.`;
    ClaseEntero.showResultado(resp);
}

File: 01 CLASE ENTERO/py/Entero.py
class Entero: 
    
    def __init__(self, Numero):
        self.Num = Numero
        
    def setNum(self):
        input_value = input("ingrese un numero: ")
        self.Num = int(input_value)
    
    def getNum(self): 
        return self.Num
    
    def showNum(self): 
        print(self.getNum())
    
    def showResultado(self, respuesta):
        print(respuesta)
    
    def incrementarNum(self): 
        print(self.Num + 1)
    
    def DecrementarNum(self):
        print(self.Num - 1)
        
    
    def esParImpar(self): 
        return (self.Num % 2 == 0)
    
    
    def esNumeroPerfecto(self):
        suma = 0
        for i in range(1, self.Num):
            if self.Num % i == 0:
                suma += i
        return suma == self.Num
    
    def esNumeroPrimo(self):
        if self.Num < 2:
            return False
        for i in range(2, int(self.Num ** 0.5) + 1):
            if self.Num % i == 0:
                return False
        return True
    
    def fibonacci(self):
        a, b = 0, 1
        while a <= self.Num:
            if a == self.Num:
                return True
            a, b = b, a + b
        return False
    
    def Pillarfibonacci(self):
        serie = [0, 1]
        while len(serie) < self.Num:
            serie.append(serie[-1] + serie[-2])
        return serie[:self.Num]
    
    def esArmstrong(self):
        num_str = str(self.Num)
        num_digits = len(num_str)
        suma = sum(int(digit) ** num_digits for digit in num_str)
        return suma == self.Num
    
    def esPerfectoGeneralizado(self, k):
        suma = 0
        for i in range(1, self.Num):
            if self.Num % i == 0:
                suma += i ** k
        return suma == self.Num
    
    def calcularFilaPascal(self):
        fila = []
        for k in range(self.Num + 1):
            fila.append(self.factorial(self.Num) // (self.factorial(k) * self.factorial(self.Num - k)))
        return fila

    # Método auxiliar para calcular el factorial
    def factorial(self, n):
        if n == 0 or n == 1:
            return 1
        resultado = 1
        for i in range(2, n + 1):
            resultado *= i
        return resultado
    
    def cifrarCesar(self, desplazamiento):
        num_str = str(self.Num)  # Convertir el número a string
        resultado = ""

        for digito in num_str:
            nuevo_digito = (int(digito) + desplazamiento) % 10  # Desplazar dentro del rango 0-9
            resultado += str(nuevo_digito)

        return resultado
    
    def tienePrimoGemelo(self):
        if not self.esNumeroPrimo():
            return False

        def es_primo(n):
            if n < 2:
                return False
            for i in range(2, int(n ** 0.5) + 1):
                if n % i == 0:
                    return False
            return True

        return es_primo(self.Num - 2) or es_primo(self.Num + 2)

    def descomponerEnFactoresPrimos(self):
        n = self.Num
        factores = {}
        divisor = 2

        while n > 1:
            potencia = 0
            while n % divisor == 0:
                n //= divisor
                potencia += 1
            if potencia > 0:
                factores[divisor] = potencia
            divisor += 1

        return factores

    def calcularCatalan(self):
        def factorial(n):
            return 1 if n == 0 else n * factorial(n - 1)

        catalan = factorial(2 * self.Num) // (factorial(self.Num + 1) * factorial(self.Num))
        return catalan

    def esMersenne(self):
        import math
        log2p1 = math.log2(self.Num + 1)
        return log2p1.is_integer() and self.esNumeroPrimo(int(log2p1))
    
    def menu(self): 
        while True: 
            print("\nMenú:")
            print("1. Establecer un nuevo número")
            print("2. Mostrar el numero actual")
            print("3. Incrementar el numero")
            print("4. Decrementar el numero")
            print("5. Verificar si es par o impar")
            print("6. Verificar si es un número perfecto")
            print("7. Verificar si es un número primo")
            print("8. Buscar en la serie de Fibonacci")
            print("9. Generar la serie de Fibonacci")
            print("10. Verificar si es un número Armstrong")
            print("11. Verificar si es un número perfecto generalizado")
            print("12. Calcular la fila N del triángulo de Pascal")
            print("13. Cifrar el número usando el cifrado César")
            print("14. Verificar si tiene primo gemelo")
            print("15. Descomponer en factores primos")
            print("16. Calcular el número de Catalan")
            print("17. Verificar si es un número de Mersenne")
            print("18. Salir")
            
            opcion = input("Ingrese una opcion: ")
            
            if opcion == "1": 
                self.setNum()
            elif opcion == "2": 
                self.showNum()
            elif opcion == "3": 
                self.incrementarNum()
            elif opcion == "4": 
                self.DecrementarNum()
            elif opcion == "5":
                if self.esParImpar(): 
                    print("El numero es par")
                else:
                    print("El numero es impar")
            elif opcion == "6":
                if self.esNumeroPerfecto():
                    print("El número es perfecto")
                else:
                    print("El número no es perfecto")
            elif opcion == "7":
                if self.esNumeroPrimo():
                    print("El número es primo")
                else:
                    print("El número no es primo")
            elif opcion == "8":
                if self.fibonacci():
                    print(f"El número {self.Num} pertenece a la serie de Fibonacci.")
                else:
                    print(f"Error: El número {self.Num} no pertenece a la serie de Fibonacci.")
            elif opcion == "9":
                print("Serie de Fibonacci hasta", self.Num, "elementos:", self.Pillarfibonacci())
            elif opcion == "10":
                if self.esArmstrong():
                    print(f"El número {self.Num} es un número de Armstrong.")
                else:
                    print(f"El número {self.Num} no es un número de Armstrong.")
            elif opcion == "11":
                k = int(input("Ingrese el valor de k: "))
                if self.esPerfectoGeneralizado(k):
                    print(f"El número {self.Num} es un número perfecto generalizado con respecto a k={k}.")
                else:
                    print(f"El número {self.Num} no es un número perfecto generalizado con respecto a k={k}.")
            elif opcion == "12":
                fila_pascal = self.calcularFilaPascal()
                print(f"Fila {self.Num} del triángulo de Pascal: {fila_pascal}")
            elif opcion == "13":
                desplazamiento = int(input("Ingrese el valor del desplazamiento: "))
                resultado_cifrado = self.cifrarCesar(desplazamiento)
                print(f"Número cifrado con desplazamiento {desplazamiento}: {resultado_cifrado}") 
            elif opcion == "14":
                if self.tienePrimoGemelo():
                    print(f"El número {self.Num} tiene un primo gemelo.")
                else:
                    print(f"El número {self.Num} no tiene un primo gemelo.")
            elif opcion == "15":
                factores = self.descomponerEnFactoresPrimos()
                print(f"Descomposición en factores primos: {factores}")
            elif opcion == "16":
                catalan = self.calcularCatalan()
                print(f"El número de Catalan para {self.Num} es: {catalan}")
            elif opcion == "17":
                if self.esMersenne():
                    print(f"El número {self.Num} es un número de Mersenne.")
                else:
                    print(f"El número {self.Num} no es un número de Mersenne.")
            elif opcion == "18":
                print("¡Hasta luego!")
                break 
            else: 
                print("Opción inválida, intente nuevamente")  
            
#Crear un objeto de la clase entero

numero = Entero(10)

#Iniciar menu 

numero.menu()